core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java

```diff
@@ -0,0 +1,330 @@
+            // which is still considered a change.
+            isChanged = true;
+            affectedEntities.add(new AffectedEntity(entity, shadowVariableReference));
+            if (oldValue != null) {
+                changedVariableNotifier.beforeVariableChanged().accept(
+                        (VariableDescriptor<Solution_>) shadowVariableReference.variableDescriptor(), entity);
+                shadowVariableReference.memberAccessor().executeSetter(entity, null);
+                changedVariableNotifier.afterVariableChanged().accept(
+                        (VariableDescriptor<Solution_>) shadowVariableReference.variableDescriptor(), entity);
+            }
+        } else {
+            var newValue = shadowVariableReference.calculator().apply(entity);
+
+            if (!Objects.equals(oldValue, newValue)) {
+                affectedEntities.add(new AffectedEntity(entity, shadowVariableReference));
+                changedVariableNotifier.beforeVariableChanged().accept(
+                        (VariableDescriptor<Solution_>) shadowVariableReference.variableDescriptor(), entity);
+                shadowVariableReference.memberAccessor().executeSetter(entity, newValue);
+                changedVariableNotifier.afterVariableChanged().accept(
+                        (VariableDescriptor<Solution_>) shadowVariableReference.variableDescriptor(), entity);
+                isChanged = true;
+            }
+        }
+        return isChanged;
+    }
+
+    record AffectedShadowVariable(int nodeId, int topologicalIndex) implements Comparable<AffectedShadowVariable> {
+        @Override
+        public int compareTo(AffectedShadowVariable heapItem) {
+            return topologicalIndex - heapItem.topologicalIndex;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (o instanceof AffectedShadowVariable other) {
+                return nodeId == other.nodeId;
+            }
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return nodeId;
+        }
+    }
+
+    private PriorityQueue<AffectedShadowVariable> createInitialChangeQueue() {
+        var heap = new PriorityQueue<AffectedShadowVariable>(instanceList.size());
+        for (var i = changed.nextSetBit(0); i >= 0; i = changed.nextSetBit(i + 1)) {
+            var topologicalOrder = graph.getTopologicalOrder(i);
+            heap.add(new AffectedShadowVariable(i, topologicalOrder));
+            if (i == Integer.MAX_VALUE) {
+                break; // or (i+1) would overflow
+            }
+        }
+        changed.clear();
+        return heap;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void updateInvalidityStatusOfAffectedEntities(Set<AffectedEntity> affectedEntities, LoopedTracker loopedTracker) {
+        for (var affectedEntity : affectedEntities) {
+            var invalidDescriptor = affectedEntity.variableUpdaterInfo.invalidityMarkerVariableDescriptor();
```

Comment:
Please look for the old name "invalidityMarker" and replace it with the new name.